### ISSUE TITLE
Task update ROLLBACK 

### ISSUE THREAD

**Issue by jvancoillie** [Guikingone/SchedulerBundle#88]
I configured a cron every minute like this `* * * * * developer /var/www/app/bin/console scheduler:consume`

I defined a `single_run` chainedTask to run related commands. 

In the logs I have `ROLLBACK` on the update of the task while it is in progress by a previous cron call. 
Could the execution of a `scheduler:consume` on tasks in progress have an impact ?

below the logs I have on the rollback

```bash
[2021-04-30T17:15:01.506813+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"deploy"} []
[2021-04-30T17:15:01.511597+00:00] app.INFO: A task has been scheduled {"task":"deploy"} []
[2021-04-30T17:15:01.513026+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"import_etablissements"} []
[2021-04-30T17:15:01.513958+00:00] app.INFO: A task has been scheduled {"task":"import_etablissements"} []
[2021-04-30T17:15:01.514673+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"import_communes"} []
[2021-04-30T17:15:01.515441+00:00] app.INFO: A task has been scheduled {"task":"import_communes"} []
[2021-04-30T17:15:01.516140+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"import_eleves"} []
[2021-04-30T17:15:01.516914+00:00] app.INFO: A task has been scheduled {"task":"import_eleves"} []
[2021-04-30T17:15:01.517692+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"import_personnels"} []
[2021-04-30T17:15:01.518456+00:00] app.INFO: A task has been scheduled {"task":"import_personnels"} []
[2021-04-30T17:15:01.523958+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-04-30T17:15:01.524758+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-04-30T17:15:01.525727+00:00] doctrine.DEBUG: SELECT t.* FROM scheduler_tasks t ORDER BY task_name ASC [] []
[2021-04-30T17:15:01.576799+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-04-30T17:15:01.583190+00:00] app.INFO: The worker has been started {"failedTasks":0,"lastExecutedTask":null} []
[2021-04-30T17:15:01.584053+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-04-30T17:15:01.584893+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-04-30T17:15:01.585379+00:00] doctrine.DEBUG: SELECT t.* FROM scheduler_tasks t ORDER BY task_name ASC [] []
[2021-04-30T17:15:01.623175+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-04-30T17:15:01.631363+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-04-30T17:15:01.655304+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-04-30T17:15:01.674965+00:00] doctrine.DEBUG: UPDATE scheduler_tasks SET body = :body WHERE task_name = :name {":name":"deploy",":body":"{\"body\":{\"tasks\":[{\"body\": [...]"} []
[2021-04-30T17:15:01.682368+00:00] doctrine.DEBUG: "ROLLBACK" [] []
[2021-04-30T17:15:01.683534+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-04-30T17:15:01.688333+00:00] doctrine.DEBUG: UPDATE scheduler_tasks SET body = :body WHERE task_name = :name {":name":"deploy",":body":"{\"body\":{\"tasks\":[{\"body\": [...]"} []
[2021-04-30T17:15:01.690061+00:00] doctrine.DEBUG: "ROLLBACK" [] []
[2021-04-30T17:15:01.691203+00:00] app.INFO: A task execution has failed {"task":"deploy"} []
[2021-04-30T17:15:01.695355+00:00] app.INFO: The worker is currently running {"failedTasks":1,"lastExecutedTask":"deploy","idle":true} []
[2021-04-30T17:15:01.695964+00:00] app.INFO: The worker has been stopped {"failedTasks":1,"lastExecutedTask":"deploy"} []
```

**Comment by jvancoillie** [comment]
I reproduced the ROLLBACK with a single task.

I suppose it is the (LOCK) option set on the first launch of `bin/console scheduler:consume` which causes the ROLLBACK on the update at the second launch of the `bin/console scheduler:consume` command.

On the second launch why does it consider that the "foo" task fails is a false positive. I will continue to investigate.

I attach a capture of the result
<img width="1821" alt="Capture d’écran 2021-05-01 à 18 58 51" src="https://user-images.githubusercontent.com/21243943/116789764-c9b54c00-aab0-11eb-94ab-5a9140e6a339.png">

**Comment by jvancoillie** [comment]
I think I identified the problem.
the first `ROLLBACK` occurs in `runPostExecutionMiddleware` for a task which is locked, which throws an exception

https://github.com/Guikingone/SchedulerBundle/blob/0a2f025eb1edbe8a1d444637177e38ca2eb055bc/src/Bridge/Doctrine/Transport/Connection.php#L218

this first exception considers that the task fails then it tries to update the task with the information of the fail and there the second `ROLLBACK` occurs.

This does not impact the expected operation in the case of doctrine thanks to the lock but it could maybe a problem on other transporters.

**Comment by jvancoillie** [comment]
maybe the task should not be updated when it is locked by `FlockStore`?

**Comment by Guikingone** [comment]
> maybe the task should not be updated when it is locked by FlockStore?

Or updated once released. 

> Or simplier not updated without changes ....

By default, tasks always have changes as we update the latest execution date.

**Comment by Guikingone** [comment]
Hi @jvancoillie 👋🏻 

I've looked at this issue this morning and I agree with this statement: 

> the first ROLLBACK occurs in runPostExecutionMiddleware for a task which is locked, which throws an exception on update

There's a problem regarding the fact that we run the middleware list for a locked task in the worker: 

```php
$this->middlewareStack->runPreExecutionMiddleware($task);

if ($lockedTask->acquire() && !$this->isRunning) {
    $this->isRunning = true;
    $this->dispatch(new WorkerRunningEvent($this));
    $this->handleTask($runner, $task);
}

$this->middlewareStack->runPostExecutionMiddleware($task);
```

IMHO, we should not run the middleware list if the task is not released, I'm gonna work on a fix 🙂

**Comment by Guikingone** [comment]
Small up @jvancoillie 👋🏻 

Could you test the fix via `dev-issue/88` when you have a moment ? Thanks 🙂

**Comment by jvancoillie** [comment]
@Guikingone 

Juste test your fix `dev-issue/88`, I fall into an infinite loop ... 😢 

```bash
[2021-05-03T15:40:15.481307+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-03T15:40:15.498043+00:00] app.INFO: A task has been scheduled {"task":"foo"} []
[2021-05-03T15:40:15.515438+00:00] php.INFO: User Deprecated: Class "Symfony\Component\HttpClient\TraceableHttpClient" should implement method "Symfony\Contracts\HttpClient\HttpClientInterface::withOptions(array $options)": Returns a new instance of the client with new default options. {"exception":"[object] (ErrorException(code: 0): User Deprecated: Class \"Symfony\\Component\\HttpClient\\TraceableHttpClient\" should implement method \"Symfony\\Contracts\\HttpClient\\HttpClientInterface::withOptions(array $options)\": Returns a new instance of the client with new default options. at /var/www/app/playground/vendor/symfony/error-handler/DebugClassLoader.php:387)"} []
[2021-05-03T15:40:15.521598+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-03T15:40:15.522995+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-03T15:40:15.525043+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t ORDER BY task_name ASC [] []
[2021-05-03T15:40:16.086330+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-03T15:40:16.102285+00:00] app.INFO: The worker has been started {"failedTasks":0,"lastExecutedTask":null} []
[2021-05-03T15:40:16.102934+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-03T15:40:16.103976+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-03T15:40:16.104997+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t ORDER BY task_name ASC [] []
[2021-05-03T15:40:16.107760+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-03T15:40:16.111560+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-03T15:40:16.118946+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-03T15:40:22.869253+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-03T15:40:22.888586+00:00] app.INFO: A task has been scheduled {"task":"foo"} []
[2021-05-03T15:40:22.918598+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-03T15:40:22.919651+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-03T15:40:22.921294+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t ORDER BY task_name ASC [] []
[2021-05-03T15:40:23.029903+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-03T15:40:23.045833+00:00] app.INFO: The worker has been started {"failedTasks":0,"lastExecutedTask":null} []
[2021-05-03T15:40:23.046964+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-03T15:40:23.048061+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-03T15:40:23.048980+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t ORDER BY task_name ASC [] []
[2021-05-03T15:40:23.052043+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-03T15:40:23.056511+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-03T15:40:23.082821+00:00] app.INFO: The task "foo" cannot be acquired [] []
[2021-05-03T15:40:23.083662+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":true} []
[2021-05-03T15:40:23.084271+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":false} []
[2021-05-03T15:40:23.084973+00:00] app.INFO: The task "foo" cannot be acquired [] []
[2021-05-03T15:40:23.085575+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":true} []
[2021-05-03T15:40:23.086255+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":false} []
[2021-05-03T15:40:23.087049+00:00] app.INFO: The task "foo" cannot be acquired [] []
[2021-05-03T15:40:23.087659+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":true} []
[2021-05-03T15:40:23.088195+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":false} []
[2021-05-03T15:40:23.088752+00:00] app.INFO: The task "foo" cannot be acquired [] []
[2021-05-03T15:40:23.089601+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":true} []
[2021-05-03T15:40:23.090245+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":false} []
[2021-05-03T15:40:23.090962+00:00] app.INFO: The task "foo" cannot be acquired [] []
[2021-05-03T15:40:23.091977+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":true} []
```

**Comment by Guikingone** [comment]
Damn ... Strange issues here 🤔

**Comment by Guikingone** [comment]
Ok, can you try again when you have a moment ? 😄

**Comment by jvancoillie** [comment]
no infinite loop anymore, but i get two `WorkerRunningEvent`for this task is that expected ?

```bash
[2021-05-03T16:20:35.452207+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-03T16:20:35.463627+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-03T16:20:35.477312+00:00] doctrine.DEBUG: INSERT INTO _symfony_scheduler_tasks (task_name, body) VALUES(:name, :body) {":name":"foo",":body":"{\"body\":{\"command\":\"app:fo [...]"} []
[2021-05-03T16:20:35.479376+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-03T16:20:35.501041+00:00] app.INFO: A task has been scheduled {"task":"foo"} []
[2021-05-03T16:20:35.546687+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-03T16:20:35.548470+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-03T16:20:35.551403+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t ORDER BY task_name ASC [] []
[2021-05-03T16:20:35.671748+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-03T16:20:35.690606+00:00] app.INFO: The worker has been started {"failedTasks":0,"lastExecutedTask":null} []
[2021-05-03T16:20:35.691357+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-03T16:20:35.692960+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-03T16:20:35.694174+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t ORDER BY task_name ASC [] []
[2021-05-03T16:20:35.697243+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-03T16:20:35.700148+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-03T16:20:35.718076+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-03T16:20:38.362410+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t WHERE t.task_name = :name LOCK IN SHARE MODE {":name":"foo"} []
[2021-05-03T16:20:38.385087+00:00] app.INFO: A task has been scheduled {"task":"foo"} []
[2021-05-03T16:20:38.424865+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-03T16:20:38.426322+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-03T16:20:38.428813+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t ORDER BY task_name ASC [] []
[2021-05-03T16:20:38.539594+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-03T16:20:38.554770+00:00] app.INFO: The worker has been started {"failedTasks":0,"lastExecutedTask":null} []
[2021-05-03T16:20:38.555409+00:00] doctrine.DEBUG: SELECT COUNT(DISTINCT t.id) FROM _symfony_scheduler_tasks t LOCK IN SHARE MODE [] []
[2021-05-03T16:20:38.556762+00:00] doctrine.DEBUG: "START TRANSACTION" [] []
[2021-05-03T16:20:38.557697+00:00] doctrine.DEBUG: SELECT t.* FROM _symfony_scheduler_tasks t ORDER BY task_name ASC [] []
[2021-05-03T16:20:38.560018+00:00] doctrine.DEBUG: "COMMIT" [] []
[2021-05-03T16:20:38.562890+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":null,"idle":false} []
[2021-05-03T16:20:38.581463+00:00] app.INFO: The worker is currently running {"failedTasks":0,"lastExecutedTask":"foo","idle":true} []
```

**Comment by jvancoillie** [comment]
before this loop fix, it looks like that it was the same issue that describe for the `single_run` task.
https://github.com/Guikingone/SchedulerBundle/issues/47#issuecomment-824139435

**Comment by Guikingone** [comment]
Thing is, if we encounter an `disabled` / `paused` task, we should not stop as we can have a new task to execute, what's the command that you're launching to consume tasks?

**Comment by jvancoillie** [comment]
> what's the command that you're launching to consume tasks?

`bin/console scheduler:consume`

**Comment by Guikingone** [comment]
Ok, so if I'm right, you're facing the following case: 

- You've launched a previous `scheduler:consume` that executed tasks without errors
- You're launching a new `scheduler:consume` that execute the same tasks

Thing is, you have a tasks which has not been fully executed during the first `scheduler:consume`?

**Comment by jvancoillie** [comment]
you're right. That's exactly it, for the test I simulate it manually but this is what happens when I have a cron which launches it  every minute.

**Comment by jvancoillie** [comment]
quick question ... why is the `WorkerRunningEvent` event dispatched twice ?
Once before looking for the runner and a second time when the runner is found ?

**Comment by Guikingone** [comment]
> Once before looking for the runner and a second time when the runner is found ?

The idea was to notify that the worker is currently doing something but you're right, it's not actually "running" 🙁

**Comment by jvancoillie** [comment]
> Ok, so if I'm right, you're facing the following case:
> 
>     * You've launched a previous `scheduler:consume` that executed tasks without errors
> 
>     * You're launching a new `scheduler:consume` that execute the same tasks
> 
> 
> Thing is, you have a task which has not been fully executed during the first `scheduler:consume`?

by studying a memory problem that I encountered on a large doctrine batch task, I noticed another problem.
when I have a single_run task which is pending in the `dueTasks` list of the first `scheduler:consume` call and on the second call it is again in the list and is executed by both calls and is therefore no longer a single_run. 
I think that when I have a little more detail on this I will open a new issue ;)

**Comment by Guikingone** [comment]
Fixed via https://github.com/Guikingone/SchedulerBundle/pull/91, planned to be released in `0.4.4`
